### ISSUE TITLE
Promote `buf beta convert` to stable

### ISSUE THREAD

**Issue by joshcarp** [bufbuild/buf#1432]
Currently this shouldn't be a problem because `buf convert` does not already exist so this just involves removing `beta`

**Comment by bufdev** [comment]
List of questions to address about current UX before we promote this to stable:

- How do well-known types work: https://github.com/bufbuild/buf/issues/1434
- What does it mean to provide a source without using the `--type` flag? From looking at the code, it seems like the `--type` flag is required, but this isn't specified in the docs.
- `--input` is meant for the encoded input, but we use "input" generally to mean the source files or Image, so this is confusing. Should address.
- If we have `-o` for output, and we keep `--input`, then `--input` should be alised to `-i`.
- Confusion/overlapping use between the `<source>` argument and the `--type` flag. There are times that `--type` means that the `<source>` argument is not used. For example, I tried to see if `buf build -o - | buf beta convert --type google.protobuf.FileDescriptorSet` would just work to address #1434, but it doesn't, you get this error message: `Failure: if a source isn't provided, the type needs to be a fully qualified path that includes the module reference; failed to parse the type: "google.protobuf.FileDescriptorSet" is not a valid fully qualified path`. There should be a clear split between source and type. Perhaps there should only be one of these.
- Documentation for `--error-format` should probably be better and more specific - in what cases does `convert` result in a build? I know the answer (when you are building the source) but we should probably be more explicit. This may be a `buf`-wide problem.
- Why does `buf beta convert` not allow you to use Images, instead only allowing you to use sources? I would think I'd be able to do `buf beta convert <(buf build -o -)#format=bin --type foo.Bar` for example.

**Comment by bufdev** [comment]
Also need better documentation when you do `buf help beta convert`.

**Comment by bufdev** [comment]
Right now, we have an `inverseEncoding` function that says "if the input is binary, output JSON, and vice versa". If we promote this to stable, this behavior will prevent us from later easily expanding to other encodings - although not impossible, the default behavior may be weirdly defined. We should make sure that users have to somehow explicitly define the input's encoding and the output's encoding. We usually do this by either file extension or options on the command line. Examples:

- `buf convert --input tmp.bin` - this would be parsed as a binary file
- `buf convert --input tmp.json` - this would be parsed as a JSON file
- `buf convert --input -` - this would default to binary
- `buf convert --input -#format=json` - this would be parsed as JSON

**Comment by joshcarp** [comment]
> Why does buf beta convert not allow you to use Images, instead only allowing you to use sources? I would think I'd be able to do buf beta convert <(buf build -o -)#format=bin --type foo.Bar for example.

Just re-using input from all other commands, no special handling. 
This just needs to be implemented yet.

> `buf beta convert <(buf build -o -)#format=bin --type foo.Bar` 

Are there any examples of this in other commands?

**Comment by joshcarp** [comment]
The term "input" means something very specific within bufland; [docs](https://docs.buf.build/reference/inputs/)

It is used to refer to "collection of .proto files" and can be a range of things;
- .proto file
- a buf module
- a built buf image


For every other command `{build,lint,breaking,generate,ls-files}` this makes sense, but as `buf beta convert` has two different inputs (the input/schema and a payload) the term `input` is now ambiguous.

There are different ways of solving this.


## 1. Keep consistent with other buf commands - "input" as first argument

```bash
buf beta convert <input> --payload=<payload> --output -#format=json
```

+ Keeps consistent with other buf commands
- Isn't too clear to what makes an "input" in this single instance
- "input" is not directly connected to "output"
- Questions arise about stdin; is it for "input" or "payload"


## 2. Break away from other commands and use the first argument as "payload"

+ This would make more sense when looking at this command in isolation
+ Less docs lookups because this particular command would be more intuitive
- Question arises to if "input" should be still called "input" and not "schema" or something

```
buf beta convert <payload> --input=foo.bin —output -#format=json

# swap "input" for "schema"

buf beta convert <payload> --schema=foo.bin —output -#format=json
```

**Comment by joshcarp** [comment]
Discussed with @cyinma we're going with number 1 because number 2 was already dismissed a couple of months ago

**Comment by joshcarp** [comment]
linked discussion: https://github.com/bufbuild/buf/pull/925#discussion_r805056791

**Comment by joshcarp** [comment]
Work that still needs to be done;
- allowing stdin redirecting into `<input>` argument and `--payload` flag of `buf beta convert` 
-

**Comment by joshcarp** [comment]
buf beta convert <(buf build -o -)#format=bin --payload <(cat payload.json)#format=json --type buf.Foo --output -#format=json is supported

**Comment by joshcarp** [comment]
Rename input -> schema or something similar

**Comment by joshcarp** [comment]
from @bufdev 
```
buf convert <PAYLOAD> -o <DEFAULTS_TO_STDOUT_JSON_FORMAT> --schema <DEFAULTS_TO_PWD> --type REQIRED
```

**Comment by joshcarp** [comment]
# Rename input -> schema or something similar

buf lint <schema>
buf breaking --against <schema>

This will also change all of the `inputs` [here](https://docs.buf.build/reference/inputs) to "schema" eventually

**Comment by joshcarp** [comment]
# Alternative names to "input"
Input is currently defined and used as "either a source—a set of .proto files—or an image—a set of .proto files built into a single, encapsulating Protobuf Image message."

Because Input is confusing seeing that more commands are being created (buf beta convert for example will use "payload" as "input" not a .proto file) the name of "input" needs to change to something more descriptive

## Schema
Schema is very well understood and recognized language; 
- protobuf schema
- avro schema
- json schema

Calling the argument `schema` instead of `input` means that the name is descriptive of what it's actually requiring. Any places where it's ambiguous can be also left as `input` as "the input to this specific command which happens to be a schema" or changed to `schema input` as opposed to `schema output` or anything else.

## Module
Module would be good language if it wasn't already used

**Comment by joshcarp** [comment]
## Schema
Schema is very well understood and recognized language; 
- protobuf schema
- avro schema
- json schema

Calling the argument `schema` instead of `input` means that the name is descriptive of what it's actually requiring. Any places where it's ambiguous can be also left as `input` as "the input to this specific command which happens to be a schema" or changed to `schema input` as opposed to `schema output` or anything else.

**Comment by joshcarp** [comment]
## Module
Because the input of commands call for different forms of proto files (either built or raw in .proto forms) the word "module" can work nicely. 

In this context "module" is also clear that it can be any module because there already is the concept of "local modules" and "remote modules" and so having the input as "module" would automatically imply that both local and remote inputs are supported without visiting the help command.


The terminology of "module" may not be desired because it implied that everything is a module (a single .proto file, for example) whereas there is no current complete definition of what is and what is not a module. 
- `buf lint <module>`
- a module is either an image or proto source or anything that resembles those things.

**Comment by joshcarp** [comment]
# object

Object does not have any existing uses within buf and could be a better alternative to `module` although it suffers from being too general just like "input";

```
buf build <object>
buf lint <object>
buf convert <payload> --object="foo.proto" // need more context to what object is
```

**Comment by joshcarp** [comment]
# Spec / Specification

Similar to `schema` but slightly different because it isn't a term that is often used with protobuf files:

```
buf build <spec> 
buf lint <spec> 
buf breaking <spec> --against=<spec-2>
buf generate <spec> 
buf ls-files <spec> 
```

**Comment by joshcarp** [comment]
PR for specifically `buf beta convert`: https://github.com/bufbuild/buf/pull/1480
PR for changing all other `buf` commands: https://github.com/bufbuild/buf/pull/1482
PR for updating documentation: https://github.com/bufbuild/docs.buf.build/pull/216

**Comment by akshayjshah** [comment]
To me, a single `.proto` file is full of schemas - it's a very general term that we'll want to use without this particular meaning. IMO, none of these are enough better than `input` to be worth the change. If we're trying to remove the overlap with `buf beta convert`, maybe we can change convert to use `--source/--dest` or `--from/--to`?

I'm fine with whatever wins the vote, though. IMO there's no perfect term.

**Comment by joshcarp** [comment]
@akshayjshah the current beta command in main follows the following format:

```
buf beta convert <input> --payload=payload.json --output=out.json
```

This keeps the same format of the other commands and also follows the same decision as [this](https://github.com/bufbuild/buf/pull/925#discussion_r805056791)

imo --from=payload.json --to=output.json works better because there's a clear symmetry between `from` and `to` which is broken with `--payload` and `--output`

```
buf beta convert <input> --from=payload.json --to=out.json
```

**Comment by joshcarp** [comment]
Fixed in #1574
